Tactics/Reflect/AxEffects.lean

```diff
@@ -0,0 +1,563 @@
+  proof : Expr
+  deriving Repr
+open AxEffects.FieldEffect
+
+/-- `AxEffects` is an axiomatic representation of effects,
+i.e., `AxEffects` transforms a description of an `ArmState` written as
+a sequence of `w` and `write_mem`s to some initial state
+into a set of hypotheses that relates reading fields from the final state
+to the initial state.
+
+`AxEffectx` contains a hashmap from `StateField` to an expression,
+in terms of the fixed initial state,
+that describes the value of the given field *after* the writes.
+Additionally, each field carries a proof that it is indeed the right value.
+
+Furthermore, we maintain a few other invariants,
+see the docstrings on each field for more detail -/
+structure AxEffects where
+  /-- The initial state -/
+  initialState : Expr
+  /-- The current state, which may be expressed in either the "exploded form"
+  (a sequence of `w`/`write_mem` to the initial state), or as just a variable
+  which is (propositionally) equal to that exploded form -/
+  currentState : Expr
+
+  /-- A map from each statefield to the corresponding `FieldEffect`.
+  If a field has no entry in this map, it has not been changed, and thus,
+  its value can be retrieved from the `nonEffectProof` -/
+  fields : Std.HashMap StateField AxEffects.FieldEffect
+  /-- An expression that contains the proof of:
+    ```lean
+    ∀ (f : StateField), f ≠ <f₁> → ⋯ → f ≠ <fₙ> →
+      r f <currentState> = r f <initialState> `
+    ```
+    where `f₁, ⋯, fₙ` are the keys of `fields`
+  -/
+  nonEffectProof : Expr
+  /-- An expression of a (potentially empty) sequence of `write_mem`s
+  to the initial state, which describes the effects on memory.
+  See `memoryEffectProof` for more detail -/
+  memoryEffect : Expr
+  /-- An expression that contains the proof of:
+    ```lean
+    ∀ n addr,
+      read_mem_bytes n addr <currentState>
+      = read_mem_bytes n addr <memoryEffect>
+    ``` -/
+  memoryEffectProof : Expr
+  /-- A proof that `<currentState>.program = <initialState>.program` -/
+  programProof : Expr
+  /-- An optional proof of `CheckSPAlignment <currentState>`.
+
+  This proof is preserved on a best-effort basis.
+  That is, if we update the state with a write to memory, or a register that is
+  not SP, we maintain the proof.
+  However, if SP is written to, no effort is made to prove alignment of the
+  new value; the field will be set to `none` -/
+  stackAlignmentProof? : Option Expr
+  deriving Repr
+
+namespace AxEffects
+
+/-! ## Initial Reflected State -/
+
+/-- An initial `AxEffects` state which has no writes.
+That is, the given `state` is assigned to be both the initial and the current
+state -/
+def initial (state : Expr) : AxEffects where
+  initialState      := state
+  currentState      := state
+  fields            := .empty
+  nonEffectProof    :=
+    mkLambda `f .default (mkConst ``StateField) <|
+      let rf := mkApp2 (mkConst ``r) (.bvar 0) state
+      mkApp2 (.const ``Eq.refl [1]) (.const ``ArmState []) rf
```

Comment:
nit: use the `mkEqArmState` helper.

Reply:
This actually needs a `mkEqReflArmState` helper, so I added that